presto-main/src/main/java/com/facebook/presto/operator/scalar/MapSubscriptOperator.java

```diff
@@ -192,88 +149,62 @@ public static Slice longSliceSubscript(Slice map, long key)
         return subscript(map, key, ExtractorType.SLICE);
     }
 
-    public static Slice longStructuralSubscript(Slice map, long key)
-    {
-        return subscript(map, key, ExtractorType.STRUCTURAL);
-    }
-
     @SuppressWarnings("unchecked")
     private static <T> T subscript(Slice map, Object key, ExtractorType type)
     {
-        JsonExtractor<?> extractor = CACHE.getUnchecked(new CacheKey(key.toString(), type));
-        return (T) JsonExtract.extract(map, extractor);
-    }
-
-    private static class CacheKey
-    {
-        private final String key;
-        private final ExtractorType type;
+        SliceInput input = map.getInput();
+        Block block = ArrayBuilder.readBlock(input);
 
-        private CacheKey(String key, ExtractorType type)
-        {
-            this.key = key;
-            this.type = type;
-        }
-
-        public String getKey()
-        {
-            return key;
-        }
-
-        public ExtractorType getType()
-        {
-            return type;
-        }
-
-        @Override
-        public boolean equals(Object o)
-        {
-            if (this == o) {
-                return true;
+        int position = 0;
+        for (; position < block.getPositionCount(); position += 2) {
+            if (block.isNull(position) && key == null) {
+                break;
             }
-            if (o == null || getClass() != o.getClass()) {
-                return false;
+            else if (key instanceof Slice && block.getSlice(position, 0, block.getLength(position)).equals(key)) {
+                break;
             }
-
-            CacheKey cacheKey = (CacheKey) o;
-
-            if (!key.equals(cacheKey.key)) {
-                return false;
+            else if (key instanceof Long && block.getLong(position, 0) == (long) key) {
```

Comment:
This doesn't quite work because of types like `timestamp with time zone` where two native values can be equal according to their native encoding, but not be equal according to Java semantics (http://prestodb.io/docs/current/develop/types.html). Instead you should bind the key type, and use `type.equalTo()` like you did in `ArrayFunctions.arrayContains()`